bios/stage-4/src/main.rs

```diff
@@ -216,6 +218,14 @@ fn init_logger(
     framebuffer_info
 }
 
+fn used_memory_slices(info: &BiosInfo) -> impl Iterator<Item = UsedMemorySlice> + Clone {
+    // skip kernel and ramdisk because they are handled individually by the
+    // uefi/bios common code
+    [info.stage_3, info.stage_4, info.config_file]
```

Comment:
None of these memory regions are needed after the kernel starts executing -> these don't have to be marked as being used by the bootloader. This also makes the `LegacyFrameAllocator::new_with_used_slices` redundant.

Reply:
I'm not 100% sure about stage 4 and the config file, but we need at least stage 3, because the gdt is part of it's data region. Otherwise the kernel might accidentally overwrite the gdt if it is not changing to a custom gdt.

Reply:
I think this is what originally made this fail:
https://github.com/rust-osdev/bootloader/pull/446#issuecomment-2158821382
> > The tests are currently failing but as far as I can tell that is due to an unrelated bug that I uncovered with this change(https://github.com/rust-osdev/bootloader/issues/445)
>
> The memory map constructed for BIOS boot is wrong. When I apply the following patch, the BIOS test crashes:

Reply:
> I'm not 100% sure about stage 4 and the config file, but we need at least stage 3, because the gdt is part of it's data region. Otherwise the kernel might accidentally overwrite the gdt if it is not changing to a custom gdt.

The common code (executed with in stage 4) [creates and loads](https://github.com/rust-osdev/bootloader/blob/3b07272dba12fb877f5679ffe07820ce53edccd8/common/src/gdt.rs#L10) a new GDT.

> I think this is what originally made this fail: [#446 (comment)](https://github.com/rust-osdev/bootloader/pull/446#issuecomment-2158821382)
> 
> > > The tests are currently failing but as far as I can tell that is due to an unrelated bug that I uncovered with this change(#445)
> > 
> > 
> > The memory map constructed for BIOS boot is wrong. When I apply the following patch, the BIOS test crashes:

The root of this failure was that memory allocated by stage-2 (e.g. kernel memory) was marked as usable. 

If I change `used_memory_slices` to return an empty iterator, all the tests still pass, so I'm reasonably confident that it's not needed.

Reply:
Maybe you are right. I still feel like overwriting the gdt should lead to issues at some point. I'm still learning a lot about this so I don't really know.
Correct me if I am wrong, but interrupts/exceptions should access the gdt. However I don't think we have any tests with interrupts enabled. 
But lets say we create a test that loads an idt with a trivial interrupt handler that just returns. Then we write all usable memory with 0. And then we trigger an interrupt. This should cause a fault or am I wrong?
I should find some time tomorrow to create said test, but I am pretty confident that this would fail.

Reply:
> Maybe you are right. I still feel like overwriting the gdt should lead to issues at some point.

That shouldn't happen though. The common code allocates memory for the GDT through `LegacyFrameAllocator` and so that memory is marked as unusable. The memory backing the GDT is never advertised as usable and so the user shouldn't overwrite it.

>  I'm still learning a lot about this so I don't really know. Correct me if I am wrong, but interrupts/exceptions should access the gdt.

Yup, you're right, `iretq` always reads from the GDT.

> However I don't think we have any tests with interrupts enabled. But lets say we create a test that loads an idt with a trivial interrupt handler that just returns. Then we write all usable memory with 0. And then we trigger an interrupt. This should cause a fault or am I wrong? I should find some time tomorrow to create said test, but I am pretty confident that this would fail.

This would cause a fault, *if* the GDT was marked as usable, but it's not.

Reply:
> That shouldn't happen though. The common code allocates memory for the GDT through LegacyFrameAllocator and so that memory is marked as unusable. The memory backing the GDT is never advertised as usable and so the user shouldn't overwrite it.

Now that you say that. I have been looking at the temporary gdt used for switching to long mode.....
Yeah your right. It should be save to delete this piece of code.